Add state measurement support for adjoint jac and adjoint vjp (#4865)
PennyLaneAI/pennylane: doc/introduction/unsupported_gradients.rst

Reviewer: unrelated but just wanted to note that I had a hard time finding this page.

Author: me too 😆

Reviewer 2: This is such a useful page, but I've never seen it before. Should we add a `.. note::` box somewhere to the Gradients and Training page that links to this page?

Reviewer 2: We could also link to it from the `diff_method` description in the QNode docstring.

Reviewer: opened #4946 to start a discussion about it without blocking this PR

PennyLane implements the adjoint differentiation method from
`2009.02823 <https://arxiv.org/pdf/2009.02823.pdf>`__, which only discusses
the gradient of expectation of observables. The implementation is specific to the paper, hence the best performance

> will occur when only using expectation values of observables.

Reviewer 2: I think this is outdated if we have support for VJPs with adjoint.

If we are supporting VJPs with adjoint, then adjoint should be efficient _as long as the final cost function that is being differentiated is a scalar_.

So a QNode with a single expectation value, _or_ a statevector/probability vector/multiple expvals that is being reduced to a scalar, etc.

This is why `<H>` is efficient, because we can reduce multiple expectation values to a scalar output.

Author: I have updated the docs to recommend using `device_vjp=True` when differentiating non-expectation values.

The challenge is *should be efficient* versus *is efficient*. I don't yet have enough numbers to promise that it does have decent performance scaling.

I have tried one case, and adjoint+device vjp seems reasonably decent for that one case, but I don't think that's enough evidence to recommend it across the board.

Reviewer 2: @albi3ro in this case though, using VJPs with backpropagation-style methods (like adjoint) will provably be better performant if the cost function being differentiated has a scalar output, compared to not using VJPs.

I understand we haven't done more formal benchmarking here, but since we are only promising 'best performance' (not efficient performance scaling, which is more rigorous), I would say that is fine. E.g., the following should be okay here:

```
The implementation is specific to the paper, hence the best performance
will occur only when the final cost function being differentiated has a
scalar value, and statevectors are not returned from QNodes.
```

Is there an underlying reason for the performance concern?

Author: I've changed the documentation here a little bit, so hopefully it's better now.

I've been running some more numbers, and getting more comfortable with recommending using adjoint diff+state, as long as `device_vjp=True`. 

My reasons for concern come back to historically seeing the difference between "should be" and "is" when it comes to performance, as implementation details can slow everything down or overheads can dominate. For example, when we starting binding device vjp's, they were substantially slower then the full jacobian path. And I have seen that the full jacobian for adjoint+state is very memory-hungry.

![scaling](https://github.com/PennyLaneAI/pennylane/assets/6364575/b1fc0a3a-28c5-429a-a96f-c1a873902b48)

>>> print_grad()
[-0.09983342]

``default.qubit`` can differentiate any other measurement process as long as it

> is in the Z measurement basis. In this case, we recommend using the device provided vjp
> ``device_vjp=True`` for improved performance scaling. ``lightning.qubit`` only supports expectation values.

Reviewer: ```suggestion
is in the Z measurement basis. In this case, we recommend using the device-provided vjp
(``device_vjp=True``) for improved performance scaling. ``lightning.qubit`` only supports expectation values.
```

.. code-block:: python 

    @qml.qnode(qml.device('default.qubit'), diff_method="adjoint", device_vjp=True)
    def circuit(x):
        qml.IsingXX(x, wires=(0,1))
        return qml.vn_entropy(wires=0)

> >>> qml.grad(circuit)(qml.numpy.array(0.1)

Reviewer: ```suggestion
>>> qml.grad(circuit)(qml.numpy.array(0.1))
```

>>> print_grad()
[-0.09983342]

> PennyLane can now use adjoint differentation for any state based measurement on `default.qubit`. `lightning.qubit`

Reviewer: ```suggestion
PennyLane can now use adjoint differentation for any state based measurement on ``default.qubit``. ``lightning.qubit``
```

> PennyLane can now use adjoint differentation for any state based measurement on ``default.qubit``. ``lightning.qubit``

Reviewer 2: ```suggestion
Adjoint differentation supports any state based measurement on ``default.qubit``. ``lightning.qubit``
```

Reviewer 2: The reason for this change is that in a sense documentation exists 'outside' of time, so 'now' is not necessary here.

still only supports expectation values.

.. code-block:: python 

    @qml.qnode(qml.device('default.qubit'), diff_method="adjoint")
    def circuit(x):
        qml.IsingXX(x, wires=(0,1))
        return qml.vn_entropy(wires=0)

>>> qml.grad(circuit)(qml.numpy.array(0.1)
(0.2989909451499196+0j)

> Furthermore, the adjoint differentiation algorithm is analytic by nature. If the user creates a device

Reviewer 2: ```suggestion
Furthermore, the adjoint differentiation algorithm is analytic by nature. If a device is created
```

Minor, but I prefer not to use 'user' in user-facing docs, sounds a bit impersonal 😆
